src/bin/pg_autoctl/watch.c

```diff
@@ -0,0 +1,1102 @@
+#include <ncurses.h>
+
+#include <sys/ioctl.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+
+#include "postgres_fe.h"
+
+#include "cli_common.h"
+#include "commandline.h"
+#include "defaults.h"
+#include "env_utils.h"
+#include "ipaddr.h"
+#include "keeper_config.h"
+#include "keeper.h"
+#include "monitor_config.h"
+#include "monitor_pg_init.h"
+#include "monitor.h"
+#include "nodestate_utils.h"
+#include "parsing.h"
+#include "pgctl.h"
+#include "pghba.h"
+#include "pgsetup.h"
+#include "pgsql.h"
+#include "pidfile.h"
+#include "state.h"
+#include "string_utils.h"
+#include "watch.h"
+#include "watch_colspecs.h"
+
+volatile sig_atomic_t window_size_changed = 0;      /* SIGWINCH */
+
+static int print_watch_header(WatchContext *context, int r);
+static int print_watch_footer(WatchContext *context);
+static int print_nodes_array(WatchContext *context, int r, int c);
+static int print_events_array(WatchContext *context, int r, int c);
+
+static ColPolicy * pick_column_policy(WatchContext *context);
+static bool compute_column_spec_lens(WatchContext *context);
+static int compute_column_size(ColumnType type, NodeAddressHeaders *headers);
+
+
+static void print_column_headers(WatchContext *context,
+								 ColPolicy *policy,
+								 int r,
+								 int c);
+
+static void print_node_state(WatchContext *context,
+							 ColPolicy *policy,
+							 int index,
+							 int r,
+							 int c);
+
+static void print_events_headers(WatchContext *context, int r, int c);
+
+static void watch_set_state_attributes(NodeState state, bool toggle);
+
+static void clear_line_at(int row);
+
+/*
+ * catch_sigwinch is registered as the SIGWINCH signal handler.
+ */
+static void
+catch_sigwinch(int sig)
+{
+	window_size_changed = 1;
+	pqsignal(sig, catch_sigwinch);
+}
+
+
+/*
+ * watch_display_window takes over the terminal window and displays the state
+ * and events in there, refreshing the output often, as when using the watch(1)
+ * command, or similar to what top(1) would be doing.
+ */
+void
+watch_main_loop(WatchContext *context)
+{
+	int step = 0;
+
+	/* the main loop */
+	for (bool first = true;; first = false)
+	{
+		instr_time start;
+		instr_time duration;
+
+		INSTR_TIME_SET_CURRENT(start);
+
+		/*
+		 * First, update the data that we want to display, and process key
+		 * strokes. We are going to update our data set twice per second, and
+		 * we want to react to key strokes and other events much faster than
+		 * this, every 50ms.
+		 */
+		step = (step + 1) % 10;
+
+		if (!watch_update(context, step))
+		{
+			/* we quit when watch_update returns false */
+			break;
+		}
+
+		/* now display the context we have */
+		if (first)
+		{
+			(void) watch_init_window(context);
+		}
+
+		(void) watch_render(context);
+
+		/* and then sleep for the rest of the 50 ms */
+		INSTR_TIME_SET_CURRENT(duration);
+		INSTR_TIME_SUBTRACT(duration, start);
+
+		int sleepMs = 50 - INSTR_TIME_GET_MILLISEC(duration);
+
+		pg_usleep(sleepMs * 1000);
```

Comment:
What happens when we have spent more than 50ms on this step? We should probably have a check for negative numbers somewhere.

Reply:
I added a protection for negative numbers, good catch.